From 81ca7f2a6badb6d559a1970ae481a4b0bcf88ab5 Mon Sep 17 00:00:00 2001
From: Guido Falsi <madpilot@FreeBSD.org>
Date: Sun, 21 Apr 2019 23:58:28 +0200
Subject: [PATCH] Bug #14857 : Fix crashing on FreeBSD

- xfburn crashing on FreeBSD when burning audio CDs (gst mutex locking issue)
---
 xfburn/xfburn-transcoder-gst.c | 16 +++++-----------
 1 file changed, 5 insertions(+), 11 deletions(-)

diff --git a/xfburn/xfburn-transcoder-gst.c b/xfburn/xfburn-transcoder-gst.c
index 7318db45..d35ed335 100644
--- a/xfburn/xfburn-transcoder-gst.c
+++ b/xfburn/xfburn-transcoder-gst.c
@@ -222,7 +222,7 @@ xfburn_transcoder_gst_init (XfburnTranscoderGst * obj)
   /* if the mutex is locked, then we're not currently seeking
    * information from gst */
   g_mutex_init (&priv->gst_mutex);
-  g_mutex_lock (&priv->gst_mutex);
+  /* Actual lock acquired in prepare() and relased in finish() */
 
   priv->discoverer = gst_discoverer_new(GST_SECOND, NULL);
 }
@@ -492,19 +492,9 @@ bus_call (GstBus *bus, GstMessage *msg, gpointer data)
 
           if (strcmp (GST_OBJECT_NAME (GST_MESSAGE_SRC (msg)), "decoder") != 0)
             break;
-          
-          if (!g_mutex_trylock (&priv->gst_mutex)) {
-            g_critical ("Lock held by another thread, can't signal transcoding start!");
-            break;
-          } else {
-#if DEBUG_GST > 0
-            DBG ("Locked mutex to signal transcoding start");
-#endif
-          }
 
           priv->gst_done = TRUE;
           g_cond_signal (&priv->gst_cond);
-          g_mutex_unlock (&priv->gst_mutex);
           break;
       } /* switch of priv->state */
 
@@ -835,6 +825,8 @@ prepare (XfburnTranscoder *trans, GError **error)
   gboolean ret;
   gint64 end_time;
 
+  g_mutex_lock(&priv->gst_mutex);
+
   priv->tracks = g_slist_reverse (priv->tracks);
 
   priv->state = XFBURN_TRANSCODER_GST_STATE_TRANSCODE_START;
@@ -933,6 +925,8 @@ finish (XfburnTranscoder *trans)
   }
   */
   recreate_pipeline (gst);
+
+  g_mutex_unlock (&priv->gst_mutex);
 }
 
 
